### PR TITLE
Climate

### PR DESCRIPTION
**Pull request by Lumpusz**
Update esphome climate schema to be in sync with HA climate schema, intended to be backward compatible, so migration can happen incrementally. This PR is part of a bundle, as aioesphomeapi, HA esphome component also needs to change. Other relate PRs:
- https://github.com/esphome/esphome/pull/1471
- https://github.com/home-assistant/core/pull/48116

### CODE REVIEW

**Comment by fraschizzato** [comment]
It has been tested? Can't see in commits/code how is that handled by esphome HA components (homeassistant/core). Expecially how it handles the deprecation warning and the actual "if not-away = home" statement in HA side (breaking change for many esphome components - ie BangBangClimate) . Also in esphome/esphome code I can't see anything related (or available for tests), can see something in midea-preset branch but not implemented coerently to aioesphomeapi.
Can't even evaluate if working/fittable/implementable. 
Knowing that aioesphomeapi is the first step of the three needed, it's also needed to see (from now) how all works togheter. I don't think is a good idea to merge a piece that - maybe - needs some adjustement to fit the other 2.

**Comment by Lumpusz** [comment]
> It has been tested? Can't see in commits/code how is that handled by esphome HA components (homeassistant/core). Expecially how it handles the deprecation warning and the actual "if not-away = home" statement in HA side (breaking change for many esphome components - ie BangBangClimate) . Also in esphome/esphome code I can't see anything related (or available for tests), can see something in midea-preset branch but not implemented coerently to aioesphomeapi.
> Can't even evaluate if working/fittable/implementable.
> Knowing that aioesphomeapi is the first step of the three needed, it's also needed to see (from now) how all works togheter. I don't think is a good idea to merge a piece that - maybe - needs some adjustement to fit the other 2.

Thanks for your input. This PR is there for the main devs to review the approach. This is indeed the first step, and if they have fundamental concerns, I would like to address them before the rest of the components are updated. Stay tuned for updates.

File: aioesphomeapi/model.py [hunk 1]
```diff
@@ -245,6 +255,13 @@ def _convert_climate_swing_modes(value):
     return [ClimateSwingMode(val) for val in value]
 
 
+def _convert_climate_presets(value):
+    return [ClimatePreset(val) for val in value]
+
+
+def _convert_string(value):
+    return [val for val in value]
```
Comment (Lumpusz) [inline aioesphomeapi/model.py:263]: @glmnet lint complains about this line, however without this, HA fails to persist the object as JSON, as this object returned from the protobuf protocol is not a list, but some kind of google iterable, so here we just convert the iterable to a standard python list, so JSON serializer can work with it. How we should solve this lint check?

Comment (glmnet) [reply aioesphomeapi/model.py:263]: https://stackoverflow.com/questions/28829236/is-it-possible-to-ignore-one-single-specific-line-with-pylint/29562085

File: setup.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
     long_description = readme_file.read()
 
 
-VERSION = '2.6.4'
+VERSION = '2.6.5'
```
Comment (glmnet) [inline setup.py:13]: version change will be on another commit

File: aioesphomeapi/model.py [hunk 3]
```diff
@@ -245,6 +255,15 @@ def _convert_climate_swing_modes(value):
...
 
 
+def _convert_climate_presets(value):
+    return [ClimatePreset(val) for val in value]
+
+
+def _convert_string(value):
+    # need to convert protobuf iterable to list, otherwise HA fails to serialize it as JSON
+    # pylint: disable=R1721
+    return [val for val in value]
```
Comment (bdraco) [inline aioesphomeapi/model.py:265]:
```suggestion
    return list(value)
```

**Comment by jesserockz** [comment]
@Lumpusz Can you update this PR?

**Comment by Lumpusz** [comment]
> @Lumpusz Can you update this PR?

Done.

File: aioesphomeapi/model.py (hunk 3 above)
Comment (jesserockz) [reply aioesphomeapi/model.py:273]: @Lumpusz Have you tried this suggestion?

Comment (Lumpusz) [reply aioesphomeapi/model.py:273]: I have not tried this, did not think it was important. This is due to the fact that lint thinks it is already a list, but it is some kind of proto specific class, and without this HA fails.

**Comment by Lumpusz** [comment]
@jesserockz When can we merge this PR? How do you plan to orchestrate the release, as the PR for HA component is still pending, this one has to go in first, increment version, and update the HA PR.

**Comment by jesserockz** [comment]
Yeah I am aware. I have it on my list to look at and test and just got busy with other stuff for the esphome beta. After I merge I will bump the version and release and then you can update the HA pr. 

Thanks

File: aioesphomeapi/model.py [hunk 4]
```diff
@@ -270,6 +289,15 @@ class ClimateInfo(EntityInfo):
     supported_swing_modes = attr.ib(
         type=List[ClimateSwingMode], converter=_convert_climate_swing_modes, factory=list
     )
+    supported_custom_fan_modes = attr.ib(
+        type=List[str], converter=_convert_string, factory=list
```
Comment (OttoWinter) [inline aioesphomeapi/model.py:293]:
```suggestion
        type=List[str], converter=list, factory=list
```

Just as an FYI - the `_convert_string` is equivalent to `list` converter.

File: aioesphomeapi/model.py [hunk 5]
```diff
@@ -289,6 +317,15 @@ class ClimateState(EntityState):
     swing_mode = attr.ib(
         type=ClimateSwingMode, converter=ClimateSwingMode, default=ClimateSwingMode.OFF
     )
+    custom_fan_mode = attr.ib(
+        type=str, default=None
+    )
+    preset = attr.ib(
+        type=ClimatePreset, converter=ClimatePreset, default=ClimatePreset.HOME
```
Comment (OttoWinter) [inline aioesphomeapi/model.py:324]:
```suggestion
        type=ClimatePreset, converter=ClimatePreset, default=ClimatePreset.ECO
```

See comment at the top of the file:
```py
# All fields in here should have defaults set
# Home Assistant depends on these fields being constructible
# with args from a previous version of Home Assistant.
# The default value should *always* be the Protobuf default value
# for a field (False, 0, empty string, enum with value 0, ...)
```

File: aioesphomeapi/model.py [hunk 6]
```diff
@@ -289,6 +317,15 @@ class ClimateState(EntityState):
     swing_mode = attr.ib(
         type=ClimateSwingMode, converter=ClimateSwingMode, default=ClimateSwingMode.OFF
     )
+    custom_fan_mode = attr.ib(
+        type=str, default=None
```
Comment (OttoWinter) [inline aioesphomeapi/model.py:321]:
```suggestion
        type=str, default=''
```

File: aioesphomeapi/model.py (hunk 5 above)
Comment (Lumpusz) [reply aioesphomeapi/model.py:324]: Shouldn't in that case HOME be the default value? HOME essentially means normal operation, a.k.a. no special preset. If we put ECO as default, what would happen with devices that don't even have ECO?

Comment (OttoWinter) [reply aioesphomeapi/model.py:324]: That depends on the HA implementation.

The main problem is: if `preset` is not set in `api_connection.cpp`, or with older esphome client versions. `aioesphomeapi` will _receive_ that as `0` aka `ECO` mode (lack of a field in protobuf means use its 0 value). `default` is not really used here and only for HA to be able to restore entity infos.

ECO as 0 value _would_ be ok if the HA implementation checks if the preset does actually exist in `self.preset_modes` ([here](https://github.com/home-assistant/core/pull/48116/files#diff-d335f2f1f824951421ef024c3be5ef2aee032239bc4bbb723b8a5b54814ae889R281).

But that's not really a clean solution (and any other api clients would have to do the same). So yes it would be best if we encode `HOME` as 0 (would be an API protocol change, but I think it's ok if we change it soon)

Comment (OttoWinter) [reply aioesphomeapi/model.py:324]: You can replicate it like this: use an old esphome version (1.18.0), or a climate device with neither `preset` nor `away` traits. Then HA will show the climate preset as `ECO`

Comment (jesserockz) [reply aioesphomeapi/model.py:324]: I too think if HOME should be default then move it to 0 on both sides. Since its not actually used yet we can safely change it on ESPHome side

Comment (Lumpusz) [reply aioesphomeapi/model.py:324]: Ok, let's do that, I will have some time to make the changes next week.

Comment (Lumpusz) [reply aioesphomeapi/model.py:386]: @jesserockz  @OttoWinter Changes to make HOME 0 are ready, please review

**Comment by jesserockz** [comment]
Replaced by #42. Thanks for keeping this going @Lumpusz
